### ISSUE TITLE
Support subscript and superscript

### ISSUE THREAD

**Issue by zipwiz** [kschroeer/rtf-html-java#2]
I am actually very happy having found your code! Unfortunately I also need subscript/superscript support because I need do transform simple technical text snippets.

I actually added "public int supersub" to RtfState, evaluating it on Control Words dn, up, sub, super and nosupersub by inserting CSS "vertical-align:sub/super; font-size: smaller;" into the SPAN.

Actually this handles only RTF sub/super/nosupersub correctly. But implementing correct handling for dn/up together with that would have been too complicated for now.

If I should find free time to make a contribution here I will certainly do so, but actually I cannot. :-/

**Comment by zipwiz** [comment]
Implementation now transforms RTF super/sub/nosupersub to HTML CSS vertical-align:sub|super; and RTF dnN/upN to HTML CSS vertical-align:(-)Npx; with CSS N being calculated from RTF N the same way as if a font-size is derived from RTF.

The last commit also fixes a flaw in calculation of N for vertical-align:Npx; : The CSS value is calculated by Math.ceil(...) which renders different "absolute result values" depending on the argument being <0 or >0. This was not recognized at first sight :-/ The test revealed the flaw, and now it is fixed.

By "spec" (of RTF), dnN/upN should not impact font-size. Instead it seems that dnN/upN simply moves the font baseline up/down by N respectively. But in the wild, dnN/upN seems to implicitly shrink the font size according to "usual expectation", that is by 2/3 or so. Therefore, the actual implementation forces a reduced font-size not only for sub/super, but for dnN/upN also. If not needed, this font size reduction should be removed for dnN/upN.

Font size reduction tries to derive the target font size from the actual font size because the subscript/superscript directive may come along with an explicit font size directive (or continue based on previous settings). If no actual font size is available because no explicit font size directiv is actually present, CSS font-size:smaller; is used. This should force the HTML renderer to reduce the font size based on the actual (HTML/CSS) font size even if we do not know the value currently.

Construction of CSS font-size directive is not "intelligent": There is actually no mechanism to check which value "should win". Instead the implemantation is based on CSS evaluation rules which state that the last value for an attribute is used if there are multiple values defined and no value "stands higher than" the other. Therefore the "basic" font-size value is written first and the "reduced" font-size value is written behind so that the "reduced font-size" wins over the "basic font-size". This way the implementation also can keep the RtfState fontSize value unchanged.

**Comment by kschroeer** [comment]
Hi Christian,

thank you for your praise and of course your help! I have merged your PR. I hope you get on with it.

**Comment by zipwiz** [comment]
Hi Kay,



thank you for merging this back!



Finally your solution is more simple than the Java Swing solution – but it
works even better for RTF parts which it does cover: Your code transforms
/line correctly and does not fill in arbitrary blank characters due to
useless use of a HTML pretty printer.



I actually have both solutions under test, and the Java Swing solution is
somewhat arbitrary in some aspects. I think the Java Swing solution has not
changed since its introduction…



However I just added 2 new issues, but I do not think that I will be able
to work on them in the near future. But these 2 are just suggestions for
improvement, which may be used as an inspiration also for other coders.



Best regards,

-- 

Christian Gosch – Systems Architecture, Project Management

inovex GmbH · Karlsruher Straße 71 · 75179 Pforzheim
+491733181016 <javascript:void(0);> · +497231319185 <javascript:void(0);> ·
christian.gosch@inovex.de · www.inovex.de
Sitz der Gesellschaft: Pforzheim · AG Mannheim, HRB 502126 ·
Geschäftsführer: Stephan Müller



*Von:* Kay Schröer <notifications@github.com>
*Gesendet:* Dienstag, 14. April 2020 14:36
*An:* kschroeer/rtf-html-java <rtf-html-java@noreply.github.com>
*Cc:* Christian Gosch <christian.gosch@inovex.de>; Author <
author@noreply.github.com>
*Betreff:* Re: [kschroeer/rtf-html-java] Support subscript and superscript
(#2)



Hi Christian,

thank you for your praise and of course your help! I have merged your PR. I
hope you get on with it.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/kschroeer/rtf-html-java/issues/2#issuecomment-613416792>,
or unsubscribe
